### ISSUE TITLE
Small changes to cleanup_schema() function

### ISSUE THREAD

**Issue by gioagu** [3dcitydb/3dcitydb#177]
Hi guys!
First off: this is not a bug, but a suggestion/request for a small enhancement.

Use case:
1) I am a (city)db superuser. I create some users with ro and rw privileges.
2) I would like to allow users with rw privileges to truncate the database (e.g. via the cleanup_schema() function)
3) To do so, I must first grant a rw-user the TRUNCATE privileges for the citydb schema(s).
4) I do it and now, the rw-user can truncate tables.

PROBLEM:

The cleanup_schema() function is written a bit "old style" (i.e. pre PostgreSQL 10) and causes the operation to fail, because the rw-user is not the owner of the sequences. This is caused by these lines:

```
  FOR rec IN 
    SELECT sequence_name FROM information_schema.sequences where sequence_schema = 'citydb'
    AND sequence_name <> 'ade_seq'
    AND sequence_name <> 'schema_seq'
  LOOP
    EXECUTE format('ALTER SEQUENCE citydb.%I RESTART', rec.sequence_name);	
  END LOOP;
```

This limitation might be overcome (at least in PostgreSQL) by dropping the above-mentioned part and slightly modifying the TRUNCATE command as follows:

`EXECUTE format('TRUNCATE TABLE citydb.%I RESTART IDENTITY CASCADE', rec.table_name);
`

Basically, you add the "RESTART IDENTITY" part, which is possible (AFAIK) already since PostgreSQL 10.

What do you think?

G

**Comment by gioagu** [comment]
Addendum:

I did some further investigation and I stumbled upon an unexpected behaviour. Or at least, it is not obvious to me reading the PostgreSQL documentation regarding the truncate.
In short: the TRUNCATE RESTART IDENTITY resets the sequences only if the corresponding table has been created with the "new" syntax:

CREATE TABLE citydb.test1(
id bigint PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
val varchar);

Otherwise it will truncate the table, but it won't reset the sequence if the table has been generated "old style" such as:

CREATE TABLE citydb.test2(
id bigserial PRIMARY KEY,
val varchar);

---------------

Small test:

DROP TABLE IF EXISTS citydb.test1;
CREATE TABLE citydb.test1(
id bigint PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
val varchar);

INSERT INTO citydb.test1 (val) VALUES
('a'),
('b'),
('c'),
('d'),
('e'),
('f');

DROP TABLE IF EXISTS citydb.test2;
CREATE TABLE citydb.test2(
id bigserial PRIMARY KEY,
val varchar);

INSERT INTO citydb.test2 (val) VALUES
('a'),
('b'),
('c'),
('d'),
('e'),
('f');

TRUNCATE citydb.test1 RESTART IDENTITY CASCADE;
TRUNCATE citydb.test2 RESTART IDENTITY CASCADE;

**Comment by clausnagel** [comment]
Hi @gioagu, thanks for reporting. Some thoughts:
- As far as I understand, PostgreSQL uses an _implicit_ sequence when `GENERATED ALWAYS AS IDENTITY` is used. However, at least for v4, we want and need _explicit_ sequences that we can query at any time to get their current and next values.
- We also want to make sure that tools can use the 3DCityDB schema in a similar way independent of whether it runs on PostgreSQL or Oracle (or maybe more databases in the future). Using implicit sequences for one database but explicit sequences for another one would violate this goal.
- We don't create the `CREATE TABLE` statements manually but use database tools for this (pgModeler and JDeveloper for v4, DbSchema for v5). We would have to find a way to configure these tools such that the statements are created as suggested. Because I would like to avoid having to postprocess the statements after they have been created by the tools.

So, my feeling is that we can't easily change to implicit sequences for v4 without a lot of side effects and impacts to existing tools. We should find another way to solve your initial issue. For v5, however, we could check whether using implicit sequences might be possible.

@yaozhihang, what are your thoughts?

**Comment by gioagu** [comment]
Hi @clausnagel,

yes, I see and agree with your remarks. One thing: as far as I can see, when you use the new syntax, you can still refer to sequences and use/manipulate them "as usual". The only point is that the name is generated automatically. If I am not mistaken, it is generally: table name + PK column name + seq.
But, yes, this would require a revision of the current software + I have no experience with Oracle. Still, I thought of sharing these findings, as I was also surprised by the (IMO) incoherent way the truncate command behaves.

FYI: for the 3DCityDB-Tools-for-QGIS I have found a workaround, so no problem. Ironically, a rw-user cannot ALTER SEQUENCE xxxxx RESTART, but indeed SELECT setval(xxxxx, ...., ...) is allowed...

G

**Comment by clausnagel** [comment]
Closing due to workaround for initial issue.
